// @jamespotz/react-simple-readmore/dist/helpers/helper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.preset = exports.animate = void 0;

var animate = function animate(_ref, fn) {
  var frameRate = _ref.frameRate,
      x = _ref.x,
      v = _ref.v,
      destX = _ref.destX,
      k = _ref.k,
      b = _ref.b,
      precision = _ref.precision,
      mass = _ref.mass,
      key = _ref.key,
      el = _ref.el;
  // x start position
  // v velocity
  var newX = x,
      newV = v;

  var animationLoop = function animationLoop() {
    // Spring stiffness, in kg / s^2
    // for animations, destX is really spring length (spring at rest). initial
    // position is considered as the stretched/compressed position of a spring
    var Fspring = -k * (newX - destX); // Damping, in kg / s

    var Fdamper = -b * newV; // Accelaration ((-k * (x-X)) + (-b * v))

    var a = (Fspring + Fdamper) / mass; // framRate of 1000/60 thats 60fps

    newV += a * (frameRate / 1000);
    newX += newV * (frameRate / 1000); // precision is when the animation should stop default is 0.01.

    if (Math.abs(newV) < precision && Math.abs(newX - destX) < precision) {
      newX = destX;
      newV = 0;
      el.style[key] = "".concat(newX, "px");
      return typeof fn === 'function' ? fn.call() : false;
    }

    if (el) el.style[key] = "".concat(newX, "px");
    requestAnimationFrame(animationLoop);
  };

  requestAnimationFrame(animationLoop);
};

exports.animate = animate;
var preset = {
  noWobble: {
    stiffness: 170,
    damping: 26
  },
  gentle: {
    stiffness: 120,
    damping: 14
  },
  wobbly: {
    stiffness: 180,
    damping: 12
  },
  stiff: {
    stiffness: 210,
    damping: 20
  },
  slow: {
    stiffness: 280,
    damping: 60
  },
  molasses: {
    stiffness: 280,
    damping: 120
  }
};
exports.preset = preset;
